### ISSUE TITLE
Should mutually recursive functions be indicated in syntax?

### ISSUE THREAD

**Issue by jamescheney** [links-lang/links#457]
Summarizing side-discussion in #455:  

Currently Links allows multiple adjacent functions to be mutually recursive ("binding groups"), as long as they aren't separated by something else (like a variable binding or typename), and it should (but currently isn't) an error to define the same function name more than once in such a group.

For example:
```
fun a() { println("a1") }
fun b() { a(); c() }
fun c() { a() }
var x = 42;
fun a() { println("a2") }
```
This has the (potential) drawback that commenting-out or rearranging code might lead to surprising errors, because.  In principle it also allows the REPL syntax to coincide with source program syntax, by allowing later bindings to shadow earlier ones; however, the above example is not writable as one "entry" in the REPL, I think this is because each REPL entry consists of either a single "var", a "fun" function group (with / without signatures), a typename, or a directive (etc.)  

[Incidentally, it's also already not the case that the REPL syntax = source file syntax even just for variable declarations
```
links> var x = 1; var y = 2;
x = 1 : Int
links> y;
<stdin>:1: Type error: Unknown variable y.
In expression: y.
```
as it appears everything after the first `;`-terminated declaration is ignored.]

The current design is intermediate between the ML-style approach (mutually-recursive bindings using "and"; later bindings shadow earlier ones; no function signatures) and the Haskell-style approach (all functions in a given module/toplevel scope are potentially mutually recursive; order doesn't matter; names have to be unique and shadowing isn't allowed; function signatures are optional).

One additional observation made in #455 is that if we were to add `and`, then it isn't clear what to do with the function signatures.  I'll give a proposal below.

This issue is to collect (further) discussion as to whether to change the current approach, and if so how.  This doesn't seem urgent but might be good to resolve before / as nontrivial changes are happening to the module system.

**Comment by SimonJF** [comment]
This is related to #359 and #412, which (as someone who is writing larger-scale Links programs and libraries), are both becoming problematic.

@dhil and I met on Monday and decided that explicit recursive functions and types as in this proposal, coupled with nominal typing for recursive types, would be very helpful in solving the two above issues. A (large) advantage of this is that we can dispense with `refineBindings` altogether.

Since this is impacting me and I'd rather see it fixed sooner rather than later, I'm planning on working on this as a priority.

**Comment by SimonJF** [comment]
I'm making some progress on this, but it's quite a big chunk of work. I've made the required syntactic changes, and eliminated the need for `refineBindings`, but am still working on unification.

Interested parties can keep up with the fun here: https://github.com/SimonJF/links/tree/remove-refinebindings.

**Comment by jamescheney** [comment]
@slindley could you please add your alternative proposal about how to indicate mutual recursion so that we can discuss it?  I think Simon is making progress on the internals and it would be good ot have agreement about what the surface syntax should be soon.

**Comment by SimonJF** [comment]
@slindley and I discussed this (amongst other things) yesterday, strangely!

@slindley 's suggestion was:
```
rec {
  fun isEven(n) { if (n == 0) { true } else { isOdd(n-1) } }
  fun isOdd(n) { if (n == 0) { false } else { isEven(n-1) } }
}
```

with the restriction that the bindings in a `rec` block must be either all types, or all functions.

I didn't like this to start, but I came round mainly because it makes adding signatures nicer. Additionally, it's similar to what Idris does (did? I've not kept up) with mutually-recursive type blocks.

It's just a few parser tweaks, so moving to this syntax shouldn't be problematic. I'd be interested to hear what others think though.

**Comment by SimonJF** [comment]
Also, interested parties might want the updated branch link: https://github.com/SimonJF/links/tree/remove-refinebindings-rebase

**Comment by dhil** [comment]
I am in favour of @slindley's proposal. Regarding the implementation, I would be liberal in the parsing phase and allow every declaration construct to occur within a `rec`-group. Then in a subsequent pass, I would check the "well-formedness" of the group.

**Comment by SimonJF** [comment]
@dhil I am now in favour, too.

Regarding your implementation suggestion -- yes, this is what I do in my current implementation (although the construct is more for functions right now). I imagine the changes I will need to make in order to adopt @slindley 's syntax will be minimal.

Does anyone have any objections?

**Comment by jamescheney** [comment]
I think this proposal needs to be fleshed out more than the above sketch.

- how does `rec {...}` make adding signatures nicer?  what problem does it solve compared to the two variants of the "and" proposal I sketched above?

- it seems undesirable (to me) to put top-level recursive functions inside a nested "block", since the contents should then be indented in order to not look funny.  This means that if we make something recursive (or not recursive) there will be a lot of whitespace-only changes.  Granted, this is also an issue with modules and other things.

- I think the keyword `rec` could be easy to miss, especially if the contents are un-indented.  Maybe `recursive` or `mutual`?  I can also imagine wanting to use `rec` as a variable name (for a record) though obviously OCaml programmers have learned to live without this.

- I sort of like the idea of allowing whatever bindings inside `rec(ursive)` and then a post-pass to decide what order things should be in (i.e., something like the current group behavior, but limited to a `rec` block).  However, presumably bindings inside nested scopes (e.g. inside the `fun` bodies, on the RHS of a `var`, etc.) will default back to sequential binding, unless there is another `rec`, right?  

- Are we going to allow `rec` to occur in an arbitrary scope or just at the toplevel? Is there a use case for having a dual keyword that introduces a sequential block inside a recursive?

- could we support both `and` (for pedantic people like me) and `rec` (for the adventurous) until everyone is happy with one or the other?

**Comment by jamescheney** [comment]
@SimonJF I'd prefer not to commit to this design until at least all of the above questions have answers.

**Comment by SimonJF** [comment]
> how does rec {...} make adding signatures nicer? what problem does it solve compared to the two variants of the "and" proposal I sketched above?

My personal feeling (although I don't really mind either way) is that because we have explicit signatures rather than annotated arguments, signatures make things messy since we need to explicitly have `fun` after `and`:
```
sig isEven : (Int) ~> Bool
fun isEven(n) { if (n == 0) { true } else { isOdd(n-1) } }
and sig isOdd : (Int) ~> Bool
fun isOdd(n) { if (n == 0) { false } else { isEven(n-1) } }
```
Note that as well as having to be explicit about whether something is a signature or a function, the `and` makes things not line up (thinking purely cosmetically here). This can be fixed by having `and` on a separate line.

> it seems undesirable (to me) to put top-level recursive functions inside a nested "block", since the contents should then be indented in order to not look funny. This means that if we make something recursive (or not recursive) there will be a lot of whitespace-only changes. Granted, this is also an issue with modules and other things.
,
This is true. Whichever proposal we go with, there will be a large ripple; this will either be due to adding `rec` and indenting, or adding `and`. Additionally, we will need to manually topo-sort definitions which aren't actually mutually-recursive (as I've had to do with the Prelude already, for example). I don't think the `rec` proposal has any real disadvantages in this dimension compared to the `and` proposal.

> I think the keyword rec could be easy to miss, especially if the contents are un-indented. Maybe recursive or mutual? I can also imagine wanting to use rec as a variable name (for a record) though obviously OCaml programmers have learned to live without this.

I think this is right, yes. Idris uses `mutual`, which is also somewhat more descriptive than just `rec` anyway. That said, `and` has required me to rename some Prelude functions already.

> I sort of like the idea of allowing whatever bindings inside rec(ursive) and then a post-pass to decide what order things should be in (i.e., something like the current group behavior, but limited to a rec block). However, presumably bindings inside nested scopes (e.g. inside the fun bodies, on the RHS of a var, etc.) will default back to sequential binding, unless there is another rec, right?

I'm a little confused here -- my implementation at the moment does not need to topo-sort mutually-recursive definitions; a function in a `rec` block can refer to any other function in the `rec` block. This does have ramifications for, say, polymorphic recursion, but I would argue having that more explicit is a good thing anyway.

> Are we going to allow rec to occur in an arbitrary scope or just at the toplevel? Is there a use case for having a dual keyword that introduces a sequential block inside a recursive?

I would be very hesitant to only allow it to occur at the toplevel. When I write Links code, I make heavy use of nested functions. Like the `and` proposal, this syntax is compositional as one would expect -- nested mutually-recursive functions can themselves contain nested mutually-recursive functions. Having a difference between toplevel and nested functions would be a restriction which would not provide any simplifications but would needlessly decrease expressiveness.

Regarding the dual keyword -- this is an interesting idea! That said, it would be difficult to implement, and I'm not entirely sure of the benefits (other than polymorphic recursion?). Are there any other languages which do something like this?

> could we support both and (for pedantic people like me) and rec (for the adventurous) until everyone is happy with one or the other?

I'm always wary of promising things regarding the parser :) But if there is substantive disagreement, then this would be an idea.

**Comment by jamescheney** [comment]
> since we need to explicitly have sig after and, but not the fun it refers to

I'm not convinced that this is required with the `and` approach: as someone (@dhil?) suggested earlier, we could permit a block of `sig`s together at the beginning of a recursive block, i.e. this:

```
sig odd : (Int) ~> Bool
sig even : (Int) ~> Bool
fun even(n) {
  if (n == 0) true
  else odd(n-1)
} and fun odd(n) {
  if (n == 0) false
  else even(n-1)
}
```
I think I prefer the strategy of keeping  signature and function definition together but am willing to live with a more flexible approach.

I gather that the claimed advantage here of `rec` is that we keep the current syntax, with alternating (optional) `sig` and `fun`.  But I'm not sure why this is dramatically better than the same code with `and` in between declarations/definitions rather than `rec{}` surrounding them.  

On reflection, I think what I object to is that, for sufficiently large functions, using `rec{}` seems to me like it will make it more difficult to locally notice when you're in the middle of a mutually recursive block, or see where it ends, since you only get a visual indication of this at the beginning (the `}` at the end is indistinguishable from any other closing brace).  The `and`s interspersed at least give a visual reminder that this function and the next one are mutually recursive.  A little more visual noise when the functions/signatures are short and clustered together seems like a price worth paying for greater readability when things are more spread out.

> mutual

Seems fine, does anyone have a pointer to documentation of how this works?  (I have a copy of the Idris book somewhere if not.)  If there is an existing design already then I'd be more comfortable borrowing it.

> my implementation at the moment does not need to topo-sort mutually-recursive definitions

My comment here was regarding @dhil 's suggestion to allow the contents of `rec` to be arbitrary declarations and post-process them later (in a way that would presumably be more restrictive than the current behavior) - presumably so that one could be more liberal in accepting e.g. a mix of recursive type declarations, function declarations, and maybe var declarations that use the functions.

```
rec {
  typename blah = ..
  typename blahblah = ...
  sig foo : blah -> Int
  fun foo (...) {...}
  fun bar : blahblah -> Int
  fun bar (...) {...}
  var xyz = foo(...)
}
```
instead of 
```
rec {
  typename blah = ..
  typename blahblah = ...
}
rec{
  sig foo : blah -> Int
  fun foo (...) {...}
  fun bar : blahblah -> Int
  fun bar (...) {...}
}
var xyz = foo(...)
```
But maybe I misinterpreted his suggestion.

> toplevel

+1 to allowing compositionally anywhere `fun` is.

> dual keyword

Just a thought, and not a particularly serious one.  But isn't wouldn't it just be syntactic sugar?
```
seq{
  fun foo() {}
  var y = ...
...
}
-->
var (foo,y,...) = (fun () {}, ..., ...)
```
(at least if `var` were allowed inside `rec` as Daniel suggests?)


>  promising things regarding the parser

I think parsing may not be the only problem with my idea... just to shoot my own idea down slightly, this would raise the issue of what if some joker does this:

```
rec {
   fun foo () {}...
   and fun bar () {} ...
# no and!
   fun baz {}
}
```
Can `baz` refer to `foo` and `bar` now? i.e. are the `and`s pointless?

But of course maybe we could also do this:

```
rec {
  rec {
  fun foo (){}...
  fun bar () {...}
  }
  fun baz() {...}
}
```
(unless we have a rule that says "no recursive immediately inside another recursive")

Anyway, I think where I'm landing is that this sounds reasonable, and if people generally prefer this approach then fine, but there are enough potential complications that I'd really like to see a clear specification of syntax and side-conditions/checks before settling on it.  Maybe this could be fleshed out as a wiki page (as a first draft of what will go into the updated documentation to describe this critical feature)?

**Comment by dhil** [comment]
> My personal feeling (although I don't really mind either way) is that because we have explicit signatures rather than annotated arguments [..]

I don't think we use signatures as an argument for one or the other. I think the current syntax for signatures is too restrictive. Currently a signature has to immediately precede its definition. I think we should be more liberal, and allow a signature (declaration) to precede its definition(s). But that is an orthogonal to the matters discussed here.

The core syntax of Links sorts of align with the syntax of JavaScript, and I think an (useless) advantage of `rec` proposal is that, one can put `rec` at the beginning of a program, and then all functions (and types) will be mutually recursive just as in JavaScript. 

As in OCaml, I think we should allow let-bindings whose right hand side is a syntactic value to appear in the `rec`-group (e.g. `let rec f = fun x -> f x`), but disallow any other let-bindings (e.g. `let rec f = if true then f else f`).

I have one question: In the presence of `rec`, what is the semantics of `fun`? Currently every function definition is implicitly recursive.

**Comment by jamescheney** [comment]
> In the presence of rec, what is the semantics of `fun`?

I guess there are two options:

1. outside of `rec`, a named `fun` is nonrecursive, i.e. such a `fun f (...) {...}` would be syntactic sugar for `var f = fun (...) {...}`, like in OCaml.  So if you want the current behavior you have to write `rec { fun (...) {...} }` or perhaps just `rec fun (...) {...}` as a convenience.
2. outside of `rec` a named `fun`can call itself, or other things already in scope, just like with `fun ` in Standard ML.

(1) seems like it would require many many changes, most of them somewhat pointless.  Links already tracks whether a named `fun` recursively calls itself (or other `wild`/recursive functions) in order to determine whether it definitely has the `wild` effect or not.  Perhaps doing (1) would make it possible to simplify this check?  (i.e. we would know syntactically whether recursion is possible or not). But this seems like a solution to an already-solved problem.  After all, we would still need to check for calls to other `wild` functions or operations such as printing, updates, communication etc.

(2) seems like the least delta to the current design and JavaScriptier.  The only objection I can think of is that the presence / absence of an explicit `rec` keyword may confuse people into thinking the semantics is (1), and hence be surprised if `fun f(...) { ... f(...)...}` is allowed without a surrounding `rec`.  (It is also the case that in this approach, `rec {fun f(...) {...} }` is equivalent to just `fun f(...) {...}`.)

Incidentally, JavaScript's scoping rules are fundamentally broken, like in R [1], Python, or classic LISP.  Functions at the top level are not mutually recursive, instead what happens is that the function names in the definitions are not "evaluated" until the function is called, with respect to whatever the dynamic environment is at the time.  In other words, rebinding variable names found inside a "recursive" function (even the function's name itself) can change the function's behavior; alpha-equivalence does not preserve meaning.  For example you can do this (say in your browser console):
```
> function fact (n) {
  if (n == 0) {
    return 1
  } else {
    return (fact(n-1) * n)
  }
}

> fact(10)
3628800
> oldfact = fact
> fact = function (n) {return (2*n)}
> oldfact(10)
180
```
Instead, if you really want "closed" recursive functions that you need to explicitly make recursive functions using your favorite recursion combinator.  I would not take it as a model to emulate.

[1] https://why-lambda.blogspot.com/2018/08/recursion-in-r.html

**Comment by SimonJF** [comment]
Firstly, it's great that we're discussing this so these issues come up and can be addressed properly.

So there are a few issues at hand here:

1. `mutual` in Idris:
See http://docs.idris-lang.org/en/latest/tutorial/typesfuns.html for the official documentation. Idris has lexically-scoped bindings except for inside `mutual` blocks. In `mutual` blocks, one can declare both types and functions. This is more important in a dependently-typed language, due to patterns such as induction-recursion.

This implementation seems to be closest to @dhil / @jamescheney's suggestion of allowing a `rec` block over arbitrary definitions.

2. Semantics of `fun` wrt. `rec` blocks:
I think this somewhat goes away if we call the `rec` block `mutual` instead (which is more accurate, in any case). I would be strongly opposed to making Links functions non-recursive by default: I don't see any advantages, and the ripple would be huge.

3. Single type of bindings / multiple types of bindings
So, the Idris way of doing things allows both functions and types in a `mutual` block. My personal preference, which I think is closer in spirit to the `and` proposal, is that we only allow a single type of binding inside a `mutual` block: we don't have dependent types, so we don't _need_ types and functions to be mutually-recursive.

To me, writing
```
mutual {
  typename blah = ..
  typename blahblah = ...
  sig foo : blah -> Int
  fun foo (...) {...}
  fun bar : blahblah -> Int
  fun bar (...) {...}
  var xyz = foo(...)
}
```

is confusing -- how can a variable be mutually defined with a function? Instead, writing

```
mutual {
  typename blah = ..
  typename blahblah = ...
}
```

and
```
mutual {
  sig foo : blah -> Int
  fun foo (...) {...}
  fun bar : blahblah -> Int
  fun bar (...) {...}
}
```

makes it precise what's going on: the first is a mutually-recursive block of types, and the second is a mutally-recursive block of functions. Having only the one kind of binding per block also solves problems with, say, directly-nested `mutual` blocks.

**Comment by jamescheney** [comment]
@SimonJF Thanks.  Just to clarify, I wasn't necessarily endorsing (or dis-endorsing) @dhil's suggestion, just trying to make sure it was clear.

In a lazy language such as Haskell, `var`-like declarations would be useful for defining lazy data structures using mutual recursion.  The case for this in a call-by-value language seems weaker, but I can think of one plausible reason to allow `var` inside `rec` or `mutual` blocks along with `fun` declarations, namely to define variables whose values are lazy data structures (where laziness is guarded explicitly using thunking / `fun() -> ...`).  For example, streams:
```
mutual {
  var evensFrom n = SCons(n, fun () {oddsFrom(n+1)}
  var oddsFrom n = SCons(n, fun () {evensFrom(n+1)}
  var numbersFromZero = evensFrom 0
} 
```
This example is slightly artificial, so isn't necessarily a reason to do it, if it isn't a use case we are particularly interested in exploring.  On the negative side, `var` is currently not recursive, i.e. you can't refer to the `var` being defined on the RHS (and I think lots of code would break otherwise).  So allowing `var`s inside `mutual` to refer to themselves would be a change in behavior, i.e. `var x = x+1` and `mutual {var x = x+1}` would have different behavior if e mentions x.

So it seems that the design decisions are:
1. `rec` vs. `mutual` - the latter seems more descriptive since it doesn't imply that other functions are non-recursive.
2. where is `rec`/`mutual` allowed - toplevel only vs any ordinary scope?  
3. what is allowed inside?  only one kind of binding (`type` or `fun`) or several?  is `rec`/`mutual` allowed immediately inside another `rec`/`mutual`?  is it an error if the contents are empty or just a single binding?  if `var` is allowed, is its definition allowed to refer to itself?

Am I missing anything?

I think my answers would be:
1. `mutual`
2. any scope
3. one kind of binding only, though I could live with allowing mixes of `var` and `fun` declarations as long as the behavior is clearly documented.

**Comment by SimonJF** [comment]
Thanks for the clarification. I don't think you're missing anything, and I'm glad we've explored the design space a bit now.

Regarding mutual `var` -- parser combinators might be another example in this space. Currently, I simply lift these to functions if necessary, which isn't so much of a problem. That said, it _is_ a pattern which arises.

My answers would be:
1. `mutual`
2. any scope
3. one kind of binding -- either all `typename`s or all `fun`s. A mutual block with a single binding in it should be at least a warning, if not an error.

**Comment by dhil** [comment]
> one kind of binding -- either all `typenames` or all `fun`s. A mutual block with a single binding in it should be at least a warning, if not an error.

Why not let type bindings and term bindings live in the same `rec` block? I don't quite understand the rationale, as I see it there would be no ambiguity since they live in two different sublanguages. I don't see the advantage of `rec` over `and typename` and `and fun` if a `rec` block can only accommodate a single kind of binding at the time.

**Comment by SimonJF** [comment]
I don't think there would be any ambiguity, no. But I don't think that's the point. Mutually-recursive functions, and mutually-recursive types, have well-defined meanings. In a dependently-typed language, there's certainly a reason to have both in the same block since types and functions can be mutually-referential. We don't have that requirement in Links.

The implementation for mixed bindings would be simple -- just split all of the typenames and functions out into two separate blocks, with typenames before the functions, and error on anything else. But I think we lose something in understandability if we allow both types of bindings in the same block. It is precisely `and typename/fun` but with different syntax -- I don't think that's necessarily a bad thing -- and it does solve the signatures issue, too.

I weight the benefits of understandability more than the extra small expressive power we gain here, but I won't die on this hill: if others are strongly in favour of mixed bindings, I'd be happy to go with that.

**Comment by dhil** [comment]
> [...] if others are strongly in favour of mixed bindings [...]

I'm strongly in favour of that. I think it is useful to be able to define together things, which logically belong together. For example if you consider a n-ary tree along with some operations on it.
```
rec {
  typename Tree(a) = [|Node:(a, Forest(a))|Empty|];
  typename Forest(a) = [|Cons:(Tree(a), Forest(a))|Nil|];

  fun mapTree(f, tr) {
    switch (tr) {
      case Empty -> Empty
      case Node(x, rest) -> Node(f(x), mapForest(f, rest))
    }
  }

  fun mapForest(f, fo) {
    switch (fo) {
      case Nil -> Nil
      case Cons(tr, rest) -> Cons(mapTree(f, tr), mapForest(f, rest))
    }
  }
}

**Comment by SimonJF** [comment]
OK -- I could be persuaded, although I'm still iffy (and I'd much rather use `mutual` than `rec` due to the issue with recursive functions).

@slindley -- given that it was your proposal, do you have any strong feelings about this discussion?

After finishing up working out how linearity fits in, this should be approaching the point where I can fix up the examples, so it'd be nice to get this decided upon.

**Comment by jstolarek** [comment]
Sorry about being late to the discussion.

Do I understand correctly that the whole problem of having to locate mutually recursive groups - either by inference or by explicit annotations - arises from the possibility of redefining bindings? Because otherwise we can just assume that everything is mutually recursive and the order of bindings won't matter. Personally, as a programmer, being able to redefine binding in the same scope is something that I abhor. I don't think I have ever faced a situation where this was a reasonable thing to do. (Shadowing bindings in a nested scope is of course a completely different thing). So I would say disallow binding redefinition in the same scope and assume everything is potentially mutually recursive (equivalent of wrapping everything in a file in a `mutual { ... }` block). REPL would require a bit of special treatment, but I think this would be an acceptable price to pay for a more elegant design.

In case the above is not a realistic proposal, I'm in favour of being able to mix functions and type definitions in one `mutual` block.

**Comment by jamescheney** [comment]
Maybe others can explain better, but here's my understanding:

- Links currently allows top-level definitions to be in any order, but adjacent function definitions are considered to form a "block".  Within a block, name-bindings are mutually recursive and need to be distinct.
- Likewise, adjacent typename definitions are also considered to be mutually recursive.
- Blocks are delimited by other definitions (e.g. a `var` would interrupt a mutually recursive function block).
- Later blocks can re-bind names bound in earlier ones, as long as separated by something so that the definitions aren't in the same block.
- Detecting and dealing with blocks is a pain, especially w.r.t. desugaring mutually recursive datatypes to mu's / aliases.
- I think we don't want Haskell-like behavior where everything in a file is implicitly mutually recursive with everything else, since that necessitates even more analysis to work out what types/functions actually mutually depend on one another.  (although I am not personally sure that this would actually break things as opposed to just being more work).  

Simon has been working on improving the last point, and making the mutual recursion explicit in the syntax helps a lot with that.

I think we could, but don't have to, adopt a rule where names in "the same" scope cannot be rebound, even if we don't adopt Haskell-style "everything is in scope of everything else" behavior.  However, given that such shadowing is conventional in ML-like languages I don't see a strong reason why this should be an error vs a warning, because it is not alpha-equivalence invariant.  Hence, a program that satisfies the restriction might step (in a rewriting-based semantics) to one that doesn't; code generation targeting Links code would have to make sure this doesn't happen; etc.

**Comment by SimonJF** [comment]
I agree with @jamescheney's analysis -- additionally, a big difference to Haskell is that we're both strict and impure, so the RHS of variable bindings are side-effecting.

> I think we could, but don't have to, adopt a rule where names in "the same" scope cannot be rebound, even if we don't adopt Haskell-style "everything is in scope of everything else" behavior.

By "scope", are we meaning a `mutual` block here? The rule I'm imagining will be that each mutually-recursive type in a block must not have the same as any other type in the mutually-recursive block, and each mutually-recursive function in a block must not have the same name as any other function in the block. I think it would be confusing (and incorrect) to have it any other way.

I use shadowing quite a lot whenever I'm writing Links or OCaml code (with variable and function names, but not so much with type names), so I'd be unwilling to give it up.

I think we are beginning to converge on a `mutual` block which can contain both types and functions?

**Comment by jamescheney** [comment]
> both strict and impure

Right, so this (and the value restriction) is why RHS's of `var` declarations inside `mutual` blocks should be syntactic values (as @dhil suggested).

> By "scope", are we meaning a mutual block here?

I meant what I thought @jstolarek meant, which would be to disallow things like this (even outside `mutual`):
```
var x = 1;
var x = 1+x;
x
```
which in Links is now legal and evaluates to 2.  In Haskell this
```
x = 1
x = 1+x
```
is illegal.

I think there is consensus that within a `mutual` block, the declared identifiers of a given sort (var/fun or typename) should be distinct.

**Comment by jamescheney** [comment]
OK, it seems to me that we're converging on the following proposal:

- `mutual` keyword (no one seemed to object)?
- allow `mutual` anywhere a `fun` or `typename` declaration is currently allowed (toplevel, inside function scope, presumably also modules etc.)
- don't allow `mutual` immediately inside another `mutual`
- allow mixture of `typename` and `fun`/`lin` definitions inside `mutual`, as well as `sig`, but not `var` (for now)
- functions outside `mutual` can still call themselves, and anything defined earlier
- retain the existing scoping behavior outside `mutual`, i.e. continue to allow re-binding the same variable/function name wherever we currently do

I know not all of these choices are uncontroversial, but overall I think this is a coherent and minimal modification, so I'd like to invoke Hoare's dictum [1] and make the smallest change that accommodates the need here, which is to facilitate cleanup of mutual recursion and datatypes, and unblock the module system work.  We can add `var` or monkey with the scoping behavior outside of `mutual` later, if we have reason to.

I suggest that we leave until end of Tuesday for comments/objections, so that @SimonJF can hopefully get his PR in order soon after.

[1] https://mail.mozilla.org/pipermail/es-discuss/2011-October/016966.html

**Comment by jstolarek** [comment]
James' argumentation convinces me, so I say 'yes' to the design he outlined.

**Comment by SimonJF** [comment]
Fixed in #528.
